topics/reference_guide/custom_language_support/language_server_protocol.md

```diff
@@ -0,0 +1,183 @@
+
+## Language Server integration
+
+Language Server is a separate process that analyzes source code and provides language-specific features to development tools.
+When creating a plugin that utilizes LSP within the IDE, there are two possibilities for providing a Language Server to end-users:
+
+- Bundle a [Language Server implementation](https://microsoft.github.io/language-server-protocol/implementors/servers/) binary as a resource delivered with a plugin.
+- Provide a possibility for users to define the location of the Language Server binary in their environment.
+
+The Prisma ORM plugin presents the first approach, which distributes the `prisma-language-server.js` script and uses a local Node.js interpreter to run it.
+
+For more complex cases, you may request to provide a detailed configuration with a dedicated [Settings](settings_guide.md) implementation.
+
+To implement a minimal but fully functional plugin, perform the following steps:
+- Within the `LspServerSupportProvider.fileOpened()` method, spin up the relevant LSP server descriptor, which can decide if the given file is supported by using the `LspServerDescriptor.isSupportedFile()` check method.
+- Tell how to start the server by implementing `LspServerDescriptor.createCommandLine()`.
+
+## Customization
+
+- To fine-tune or disable the implementation of LSP-based features, the plugins may override the corresponding properties of the `LspServerDescriptor` class.
+  See the properties documentation for more details:
+  - `lspGoToDefinitionSupport`
+  - `lspCompletionSupport`
+  - `lspDiagnosticsSupport`
+  - `lspCodeActionsSupport`
+  - `lspCommandsSupport`
+- To handle custom (undocumented) requests and notifications from the LSP server, override `LspServerDescriptor.createLsp4jClient`.
+- To send custom (undocumented) requests and notifications to the LSP server, override `LspServerDescriptor.lsp4jServerClass` and implement the `LspClientNotification` and/or `LspRequest` classes.
+
+See bundled LSP API source code and its code documentation for more information.
+
+## Testing
+
+Utility classes used for testing are not yet extracted for public use, but we'll provide suitable methods for testing LSP-based third-party plugins as soon as possible.
+
+## Troubleshooting
+
+All the IDE and LSP server communication logs are passed to the IDE log file.
+
+To include them for preview, add the following entry to the <control>Help | Diagnostic Tools | Debug Log Settings…</control> configuration dialog:
+
+```
+#com.intellij.platform.lsp
+```
+
+For more information, see the [](ide_infrastructure.md#logging) section.
+
+## Limitations
+
+- The current LSP API implementation assumes that the IDE <-> LSP server communication channel is `stdio`.
+- The IDE doesn't send [workspace/didChangeWatchedFiles](https://microsoft.github.io/language-server-protocol/specification/#workspace_didChangeWatchedFiles) notifications to the server.
+
+## Conclusion
```

Comment:
IMHO this is not a _Conclusion_ section, but a regular one, which gives tips about what should be considered when implementing LSP support. I would rename it to _Considerations_ or something better.